overlord/snapshotstate/snapshotstate.go

```diff
@@ -541,9 +568,49 @@ func Forget(st *state.State, setID uint64, snapNames []string) (snapsFound []str
 	return summaries.snapNames(), ts, nil
 }
 
+// setSnapshotOpInProgress marks the given set ID as being a subject of
+// snapshot op inside state cache. The state must be locked by the caller.
+func setSnapshotOpInProgress(st *state.State, setID uint64, op string) {
+	var snapshotOps map[uint64]string
+	if val := st.Cached("snapshot-ops"); val != nil {
+		snapshotOps, _ = val.(map[uint64]string)
+	} else {
+		snapshotOps = make(map[uint64]string)
+	}
+	snapshotOps[setID] = op
+	st.Cache("snapshot-ops", snapshotOps)
+}
+
+// UnsetSnapshotOpInProgress un-sets the given set ID as being a
+// subject of a snapshot op. It returns the last operation (or empty string
+// if no op was marked active).
+// The state must be locked by the caller.
+func UnsetSnapshotOpInProgress(st *state.State, setID uint64) string {
+	var op string
+	if val := st.Cached("snapshot-ops"); val != nil {
+		var snapshotOps map[uint64]string
+		snapshotOps, _ = val.(map[uint64]string)
+		op = snapshotOps[setID]
+		delete(snapshotOps, setID)
+		st.Cache("snapshot-ops", snapshotOps)
+	}
+	return op
+}
+
 // Export exports a given snapshot ID
-// Note that the state much be locked by the caller.
-var Export = backend.NewSnapshotExport
+// Note that the state must be locked by the caller.
+func Export(ctx context.Context, st *state.State, setID uint64) (se *backend.SnapshotExport, err error) {
+	if err := checkSnapshotConflict(st, setID, "forget-snapshot"); err != nil {
+		return nil, err
+	}
+
+	setSnapshotOpInProgress(st, setID, "export-snapshot")
+	se, err = backendNewSnapshotExport(ctx, setID)
+	if err != nil {
+		UnsetSnapshotOpInProgress(st, setID)
```

Comment:
(nitpick) Should we log here if we get an empty return from UnsetSnapshotOpInProgress() here? AIUI we expect something here because of the setSnapshotOpInProgress() ?

Reply:
Hmm, I don't think so, we would always get what we just set above (if not then this would be a bug in set..).